toml/src/main/kotlin/org/rust/toml/crates/local/CratesLocalIndexService.kt

```diff
@@ -0,0 +1,236 @@
+import com.intellij.openapi.components.State
+import com.intellij.openapi.components.Storage
+import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.progress.Task
+import com.intellij.util.io.*
+import org.eclipse.jgit.lib.ObjectId
+import org.eclipse.jgit.lib.Repository
+import org.eclipse.jgit.revwalk.RevTree
+import org.eclipse.jgit.revwalk.RevWalk
+import org.eclipse.jgit.storage.file.FileRepositoryBuilder
+import org.eclipse.jgit.treewalk.TreeWalk
+import org.rust.openapiext.pluginDirInSystem
+import org.rust.stdext.cleanDirectory
+import org.rust.toml.crates.local.CratesLocalIndexService.Companion.CratesLocalIndexState
+import java.io.BufferedReader
+import java.io.DataInput
+import java.io.DataOutput
+import java.io.InputStreamReader
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.util.concurrent.atomic.AtomicReference
+
+@State(name = "CratesLocalIndexState", storages = [Storage("rust.crateslocalindex.xml")])
+class CratesLocalIndexService : PersistentStateComponent<CratesLocalIndexState>, Disposable {
+    private val baseCratesLocalRegistryDir: Path
+        get() = pluginDirInSystem().resolve("crates-local-index")
+
+    private val corruptionMarkerFile: Path
+        get() = baseCratesLocalRegistryDir.resolve(CORRUPTION_MARKER_NAME)
+
+    private val userCargoIndexDir: Path
+        get() = Paths.get(System.getProperty("user.home"), CARGO_REGISTRY_INDEX_LOCATION)
+
+    private val repository: Repository = FileRepositoryBuilder()
+        .setGitDir(userCargoIndexDir.toFile())
+        .build()
+
+    private val registryHeadCommitHash: String? = repository.resolve("origin/master")?.name
+
+    private val crates: PersistentHashMap<String, CargoRegistryCrate> = run {
+        resetIndexIfNeeded()
+
+        val file = baseCratesLocalRegistryDir.resolve("crates-local-index")
+
+        IOUtil.openCleanOrResetBroken({
+            PersistentHashMap(
+                file,
+                EnumeratorStringDescriptor.INSTANCE,
+                CrateExternalizer,
+                4 * 1024,
+                CRATES_INDEX_VERSION
+            )
+        }, file)
+    }
+
+    private val _state: AtomicReference<CratesLocalIndexState> = AtomicReference(CratesLocalIndexState())
+
+    @Volatile
+    var isReady: Boolean = true
+
+    override fun getState(): CratesLocalIndexState = _state.get()
+    override fun loadState(state: CratesLocalIndexState) {
+        _state.set(state)
+    }
+
+    private fun updateIfNeeded() {
+        if (state.indexedCommitHash != registryHeadCommitHash && isReady) {
+            isReady = false
+            CratesLocalIndexUpdateTask().queue()
+        }
+    }
+
+    private fun resetIndexIfNeeded() {
+        if (corruptionMarkerFile.exists()) resetIndex()
+    }
+
+    private fun resetIndex() {
+        baseCratesLocalRegistryDir.cleanDirectory()
+        state.indexedCommitHash = ""
```

Comment:
```suggestion
state = CratesLocalIndexState()
```
since `indexedCommitHash` field is not volatile (but `state` is volatile)